doc/user-manual/language/cubical.lagda.rst

```diff
@@ -1,106 +1,930 @@
+
+As everything has to work up to higher dimensions the Glue types take
+a partial family of types that are equivalent to the base type ``A``:
+
+::
+
+  Glue : ∀ {ℓ ℓ'} (A : Set ℓ) {φ : I}
+       → Partial φ (Σ[ T ∈ Set ℓ' ] T ≃ A) → Set ℓ'
+
+..
+  ::
+
+  Glue A Te = primGlue A (λ x → Te x .fst) (λ x → Te x .snd)
+
+These come with a constructor and eliminator:
+
+.. code-block:: agda
+
+  glue : ∀ {ℓ ℓ'} {A : Set ℓ} {φ : I} {Te : Partial φ (Σ[ T ∈ Set ℓ' ] T ≃ A)}
+       → PartialP φ T → A → Glue A Te
+
+  unglue : ∀ {ℓ ℓ'} {A : Set ℓ} (φ : I) {Te : Partial φ (Σ[ T ∈ Set ℓ' ] T ≃ A)}
+         → Glue A Te → A
+
+Using Glue types we can turn an equivalence of types into a path as
+follows:
+
+::
+
+  ua : ∀ {ℓ} {A B : Set ℓ} → A ≃ B → A ≡ B
+  ua {_} {A} {B} e i = Glue B (λ { (i = i0) → (A , e)
+                                 ; (i = i1) → (B , idEquiv B) })
+
+The idea is that we glue on ``A`` to ``B`` when ``i = i0`` using ``e``
+and ``B`` to itself when ``i = i1`` using the identity
+equivalence. This hence gives us the key part of univalence: a
+function for turning equivalences into paths. The other part of
+univalence is that this map itself is an equivalence which follows
+from the computation rule for ``ua``:
+
+::
+
+  uaβ : ∀ {ℓ} {A B : Set ℓ} (e : A ≃ B) (x : A) → transport (ua e) x ≡ e .fst x
+  uaβ e x = transportRefl (e .fst x)
+
+Transporting along the path that we get from applying ``ua`` to an
+equivalence is hence the same as applying the equivalence. This is
+what makes it possible to use the univalence axiom computationally in
+Cubical Agda: we can package up our equivalences as paths, do equality
+reasoning using these paths, and in the end transport along the paths
+in order to compute with the equivalences.
+
+For more results about Glue types and univalence see
+https://github.com/agda/cubical/blob/master/Cubical/Core/Glue.agda and
+https://github.com/agda/cubical/blob/master/Cubical/Foundations/Univalence.agda. For
+some examples of what can be done with this for working with binary
+and unary numbers see
+https://github.com/agda/cubical/blob/master/Cubical/Data/BinNat/BinNat.agda.
+
+
+Higher inductive types
+----------------------
+
+Cubical Agda also lets us directly define higher inductive types as
+datatypes with path constructors. For example the circle and `torus
+<https://en.wikipedia.org/wiki/Torus>`_ can be defined as:
+
+::
+
+  data S¹ : Set where
+    base : S¹
+    loop : base ≡ base
+
+  data Torus : Set where
+    point : Torus
+    line1 : point ≡ point
+    line2 : point ≡ point
+    square : PathP (λ i → line1 i ≡ line1 i) line2 line2
+
+Functions out of higher inductive types can then be defined using
+pattern-matching:
+
+::
+
+  t2c : Torus → S¹ × S¹
+  t2c point        = ( base , base )
+  t2c (line1 i)    = ( loop i , base )
+  t2c (line2 j)    = ( base , loop j )
+  t2c (square i j) = ( loop i , loop j )
+
```

Comment:
On the right-hand-sides, the spaces after the opening parentheses and before the closing ones look a bit weird, is there a reason to have them?

Reply:
Fixed, now the result of t2c looks exactly like the argument to c2t